kitchen sink 11ty config, plus comments.

warpfork opened the pull request:
I don't know how feasible it is to make a copypasta of "how to start a site, with most of the plugins you probably want", but this is an attempt.  (Not all features are immediately used.)

Ideally it becomes organized in such a way that you can comment out full hunks of adjacent code to remove a feature, and that works; or conversely, add a single hunk of code in one contiguous text change, and have that result in a completely introduced feature.  (I'd like to have a reference template for this because I foresee doing more than one site like this in the fairly near future.)

That goal is not quite recognized yet.
Several functions reach across multiple uses in a way that confounds it; these are probably fixable (and some comments point the probable way). Also problematic is that some features are interacting with layout templates in a way that's contained in other files, far away; figuring out a way we could package these together is a bit beyond what I've had time to contemplate so far.
Still, I'd like to aim in that general direction.

This is a checkpoint, and fully tested; it may need some amending.

warpfork committed:
kitchen sink 11ty config, plus comments.

I don't know how feasible it is to make a copypasta of
"how to start a site, with most of the plugins you probably want",
but this is my attempt.

Ideally it becomes organized in such a way that you can comment out
full hunks of adjacent code to remove a feature, and that works;
or conversely, add a single hunk of code in one contiguous text change,
and have that result in a completely introduced feature.

That goal is not quite recognized yet.
Several functions reach across multiple uses in a way that confounds it;
these are probably fixable (and some comments point the probable way).
Also problematic is that some features are interacting with layout
templates in a way that's contained in other files, far away;
figuring out a way we could package these together is a bit beyond
what I've had time to contemplate so far.
Still, I'd like to aim in that general direction.

This is a checkpoint, and fully tested; it may need some amending.

chunkydotdev commented on .eleventy.js:
```diff
@@ -1,7 +1,141 @@
+
+	return {
+		dir: {
+			input:  "src",
```
issue here is that we need the README.md to have a copy in the /src folder right? 🤔  we could make a copy-script that runs before builds and creates a `src/index.md`, what do you think of that?

chunkydotdev replied:
Or does the regexp in the start of the file fix that?

warpfork replied:
Yeah.  This is logically the same as before the PR (although it renders as diff, because formatting and all the neighboring changes).  But, yes, the readme and the index are pretty much repeated.

I do sort of which we could fix that duplication problem.

I think it may be possible to set the input dir to ".".  Or, I even tried some fancy footwork in another repo previously where I put the whole 11ty config in a deeper dir, and configured `input:".."` (and it worked... although with some fun debugging, to get it started).

I'm a little torn on which angle of approach makes for the least work.

warpfork replied:
Now that I think about it a bit longer, I also do recall that github processes `/absolute/to/site/root` links as relatively the repo root in the way they render markdown docs, so, having the input path root be the same as the repo root has some nice advantage if we want to keep the links in the source material working in github's rendering as well as in the generated site output.

chunkydotdev replied:
Yeah, that would be cool, we can start like this and create an issue of that which we can fix in a future PR?

warpfork replied:
Issue made!

chunkydotdev commented on .eleventy.js:
```diff
@@ -1,7 +1,141 @@
+		eleventyConfig.addFilter("cachebuster", () => buildtime)
+	}()
+
+	// Here are all the additional files and folders unique to this site.
```
Nice! We could also lift this out of the main function and name it to make it obvious what it's used for, but I guess it works with adding a comment on top like you've done as well.

warpfork replied:
Agree, I think that could be a useful subsequent step.

chunkydotdev commented on .eleventy.js:
```diff
@@ -1,7 +1,141 @@
+			permalinkSpace: false, // Again, please don"t add actual text to the inside of the hN tag.
+			permalinkBefore: true,
+			// Generate anchors for everything but h1.  h1 tags are for page titles, and are generally not useful to jump to, so don't make anchors for those.
+			level: [2, 3, 4, 5, 6],
```
I find it kind of confusing that headings are automatically anchor-tags, I guess it's nice if you want to copy a link which you can do by right-clicking it. But design-wise I would rather have an anchor-icon next to it that has the link which you can copy.

warpfork replied:
You do like the anchor icon to appear even next to the h1 with the page title in it, you mean?

chunkydotdev replied:
Yes, that would be nice! Instead of the title appearing as a link

chunkydotdev replied:
But we can add that in a future PR

chunkydotdev commented on .eleventy.js:
```diff
@@ -1,7 +1,141 @@
+	return {
+		dir: {
+			input:  "src",
+			output: "_public"
```
_public should be added to .gitignore

warpfork replied:
fixed!

chunkydotdev reviewed (changes requested):
Ah dangit, forgot to select "Request changes" 😅

warpfork committed:
rediscovered that markdown-it-replace-link is used; it's just spooky action at a distance.

warpfork committed:
fix gitignore to match new config, and also use rooted paths.

warpfork committed:
update npm package files to match new config's demands.

warpfork committed:
use rooted links.

This is generally my preference when doing linking within a site,
because it's the easiest to update page locations with a regexp.

Alternatively, switching these to "./" would've worked (and this
site is so simple it really makes no matter).
However, either that change or this one would've been required...
because our replaceLink feature doesn't engage unless there's at
least one slash.  Ahheh.

(I would really like 11ty to get a core feature for this link fixup
stuff.  It's kinda core to 99% of usage, imo.)

warpfork commented:
Okay, I fixed _many_ things.   Embarrassed to confess I hadn't run my first pushed code nearly enough >.<'  (Will be good to get CI stuff enabled for this soon maybe...)

I think this is viable to merge now.

chunkydotdev commented:
LGTM!
